templates/refresh_facts.erb

```diff
@@ -11,10 +11,31 @@ opt.on("--out FILE", "-o", "Path to facts") do |v|
   @outfile = v
 end
 
+opt.on("--pid FILE", "-p", "Path to pidfile") do |v|
+  @pid = v
+end
+
 opt.parse!
 
 abort("Please specify a file to write to") unless @outfile
 
+if @pid
+  begin
+    lock = File.open(@pid, File::CREAT | File::EXCL)
```

Comment:
you have to open this for writing else it can't write and your `at_exit` doesn't get installed so it kills the whole thing.

lets add the `at_exit` before the `write` line to give it maximal chance of working.

Also update the abort message like this:

```
abort("Failed to obtain lock for %s using '%s', the refresher may already be running: %s: %s" % [$0, @pid, $!.class, $!.to_s])
```

This makes it easier to debug things should it go south